### ISSUE TITLE
undefined method 'to_hash' when compat mode

### ISSUE THREAD

**Issue by fahchen** [ohler55/oj#315]
```ruby
class A < BasicObject
  def initialize(data)
    @data = data
  end

  def to_json
    @data
  end
end

data = Oj.dump({ ary: [1, 2] }) #=> "{\"ary\":[1,2]}"
a = A.new(data)
h = { obj: a }

json = Oj.dump(h, mode: :compat, use_to_json: true)
#=> NoMethodError: undefined method `to_hash' for #<A:0x007fae03de1dc8>
```
what I expected is `"{\"obj\":{\"ary\":[1,2]}}"`

**Comment by ohler55** [comment]
You are correct it does not work. You took a few extra steps. It would have been sufficient to use
```ruby
data = A.new('foo')
```
There is a bug where Oj expects rb_respond_to() true or false when the actual return is Qtrue and Qfalse. Good catch.

**Comment by ohler55** [comment]
Dug deeper and found the source for Ruby I looked at on the web did not match what I observed in running the code. Bottom line is that rb_respond_to() is returning 1 instead of 0 for the to_hash() method. Since the use_to_json encompasses to_json, as_json, and to_hash rb_respond_to returning non-zero causes to_hash() to be called. There is not much Oj can do if the underlying ruby returns the wrong value. I'll check some of the older versions and see if it happens there too.

**Comment by fahchen** [comment]
Cause `A` is inherited from `BasicObject`, `A` is not implement with `respond_to?`

**Comment by ohler55** [comment]
That is correct. The error raised should indicate respond_to? is not defined and never get to the call to to_hash.

**Comment by ohler55** [comment]
Using BasicObject gives up all the core functionality expected of Ruby objects.

**Comment by fahchen** [comment]
I fix it by adding `respond_to?` method to `A`
```ruby
class A < BasicObject
  def initialize(data)
    @data = data
  end

  def to_json
    @data
  end

  def respond_to?(symbol, include_all=false)
    if symbol === :to_json
      true
    else
      false
    end
  end
end

data = Oj.dump({ ary: [1, 2] })
a = A.new(data)
h = { obj: a }

json = Oj.dump(h, indent: 2, mode: :compat, use_to_json: true)
#=> "{\n  \"obj\":{\":ary\":[1,2]}\n}\n"
```

**Comment by ohler55** [comment]
Right. That is a fix although it hides the problem with Ruby 2.3.x.

**Comment by fahchen** [comment]
I got it, thanks for your help.

**Comment by ohler55** [comment]
We all work together. I submitted the Ruby issue due to your discovery. It make Ruby and Oj better when bugs are reported. Thank you.

**Comment by libc** [comment]
I and @timflapper did a bit of investigation into it yesterday.

The bug is caused by the [this revert](https://github.com/ruby/ruby/commit/53369d7e4dab675ede11b584ffc003e51ea62672) and is fixed in trunk and ruby 2.4.0 by [this commit](https://github.com/ruby/ruby/commit/b51174a6b0997467d02242d6ca5dd08c43d45fa8).

Hope that helps.

**Comment by ohler55** [comment]
Thank you.
